ygot/struct_validation_map.go

```diff
@@ -319,3 +321,190 @@ func MergeJSON(a, b map[string]interface{}) (map[string]interface{}, error) {
+func copyPtrField(dstField, srcField reflect.Value) error {
+	if srcField.IsNil() || !srcField.IsValid() {
+		return nil
+	}
+
+	// Check for struct ptr, or ptr to avoid panic.
+	if util.IsValueStructPtr(srcField) {
+		var d reflect.Value
+		d = reflect.New(srcField.Type().Elem())
+		if err := copyStruct(d.Elem(), srcField.Elem()); err != nil {
+			return err
+		}
+		dstField.Set(d)
+		return nil
+	}
+
+	if util.IsValuePtr(srcField) {
+		p := reflect.New(srcField.Type().Elem())
+		p.Elem().Set(srcField.Elem())
+		dstField.Set(p)
+		return nil
+	}
+
+	return fmt.Errorf("received non-ptr type: %v", srcField.Kind())
+}
+
+// copyInterfaceField copies srcField into dstField. Both srcField and dstField
+// are reflect.Value structs which contain an interface value.
+func copyInterfaceField(dstField, srcField reflect.Value) error {
+	if srcField.IsNil() || !srcField.IsValid() {
+		return nil
+	}
+
+	if !util.IsValueInterface(srcField) || !util.IsValueStructPtr(srcField.Elem()) {
+		return fmt.Errorf("invalid interface type received: %T", srcField.Interface())
+	}
+
+	s := srcField.Elem().Elem() // Dereference to a struct.
+	var d reflect.Value
+	d = reflect.New(s.Type())
+	if err := copyStruct(d.Elem(), s); err != nil {
+		return err
+	}
+	dstField.Set(d)
+	return nil
+}
+
+// copyMapField copies srcField into dstField. Both srcField and dstField are
+// reflect.Value structs which contain a map value. If dstField is populated
+// then an error is returned.
+// TODO(robjs): Implement merging of maps when they are populated in the dstField
+// supplied. See https://github.com/openconfig/ygot/issues/74.
+func copyMapField(dstField, srcField reflect.Value) error {
+	if !util.IsValueMap(srcField) {
+		return fmt.Errorf("received a non-map type in copy map field: %v", srcField.Kind())
+	}
+
+	if srcField.Len() == 0 {
+		return nil
+	}
+
+	if dstField.Len() != 0 {
+		return fmt.Errorf("unimplemented: cannot map slice where destination was set: %v == %v", srcField.Type().Name(), dstField.Interface())
+	}
+
+	keys := srcField.MapKeys()
+	if k := srcField.MapIndex(keys[0]).Kind(); k != reflect.Ptr {
+		return fmt.Errorf("invalid map, got member type %s", k)
+	}
+
+	// TODO(robjs): When we move to go1.9+ only support in ygot, we can use the
+	// following to make a map of a particular size.
+	// nm := reflect.MakeMapWithSize(reflect.MapOf((keys[0]).Type(), srcField.MapIndex(keys[0]).Type()), srcField.Len())
+	nm := reflect.MakeMap(reflect.MapOf((keys[0]).Type(), srcField.MapIndex(keys[0]).Type()))
+	for _, k := range keys {
+		v := srcField.MapIndex(k)
+		d := reflect.New(v.Elem().Type())
+		if err := copyStruct(d.Elem(), v.Elem()); err != nil {
+			return err
+		}
+		nm.SetMapIndex(k, d)
+	}
+	dstField.Set(nm)
+	return nil
+}
+
+// copySliceField copies srcField into dstField. Both srcField and dstField are
+// reflect.Value structs which represent slices. If the slice in dstField is populated
+// an error is returned.
```

Comment:
Both srcField and dstField are reflect.Value structs which represent slices ->
Both srcField and dstField must be reflect.Slice kind and contain struct ptr elements?

Reply:
Updated with similar wording.